lib/ansible/modules/cloud/amazon/s3_sync.py

```diff
@@ -0,0 +1,492 @@
+                'chopped_path':chopped_path,
+                'modified_epoch': f_modified_epoch,
+                'bytes': f_size
+        })
+        # dirpath = path *to* the directory
+        # dirnames = subdirs *in* our directory
+        # filenames
+    return ret
+
+def calculate_s3_path(filelist, key_prefix=''):
+    ret = []
+    for fileentry in filelist:
+        # don't modify the input dict
+        retentry = fileentry.copy()
+        retentry['s3_path'] = os.path.join(key_prefix, fileentry['chopped_path'])
+        ret.append(retentry)
+    return ret
+
+def calculate_local_etag(filelist, key_prefix=''):
+    '''Really, "calculate md5", but since AWS uses their own format, we'll just call
+       it a "local etag". TODO optimization: only calculate if remote key exists.'''
+    ret = []
+    for fileentry in filelist:
+        # don't modify the input dict
+        retentry = fileentry.copy()
+        retentry['local_etag'] = calculate_multipart_etag(fileentry['fullpath'])
+        ret.append(retentry)
+    return ret
+
+def determine_mimetypes(filelist, override_map):
+    ret = []
+    for fileentry in filelist:
+        retentry = fileentry.copy()
+        localfile = fileentry['fullpath']
+
+        # reminder: file extension is '.txt', not 'txt'.
+        _, file_extension = os.path.splitext(localfile)
+        if override_map and override_map.get(file_extension):
+            # override? use it.
+            retentry['mime_type'] = override_map[file_extension]
+        else:
+            # else sniff it
+            retentry['mime_type'], retentry['encoding'] = mimetypes.guess_type(localfile, strict=False)
+
+        # might be None or '' from one of the above. Not a great type but better than nothing.
+        if not retentry['mime_type']:
+            retentry['mime_type'] = 'application/octet-stream'
+
+        ret.append(retentry)
+
+    return ret
+
+def head_s3(s3, bucket, s3keys):
+    retkeys = []
+    for entry in s3keys:
+        retentry = entry.copy()
+        # don't modify the input dict
+        try:
+            retentry['s3_head'] = s3.head_object(Bucket=bucket, Key=entry['s3_path'])
+        except botocore.exceptions.ClientError as err:
+            if hasattr(err, 'response') and 'ResponseMetadata' in err.response and 'HTTPStatusCode' in err.response['ResponseMetadata'] and str(err.response['ResponseMetadata']['HTTPStatusCode']) == '404':
+                pass
+            else:
+                raise Exception(err)
+            #error_msg = boto_exception(err)
+            #return {'error': error_msg}
+        retkeys.append(retentry) 
+    return retkeys
+
+def filter_list(s3, bucket, s3filelist, strategy):
+    keeplist = list(s3filelist)
+
+    for e in keeplist:
+      e['_strategy'] = strategy
+
+    # fetch info from S3 if we're going to use it for comparisons
+    if not strategy == 'force':
+        keeplist = head_s3(s3, bucket, s3filelist)
+
+    if strategy == 'checksum':
```

Comment:
it's not meant to be an `elif`; I'm updating comments, though. The `not` is basically a common section so it isn't duplicated in the `if/elif/elif` strategy section.